### DISCUSSION TITLE
Entity gets unloaded, even when it's referenced with 'back_populates'

### DISCUSSION THREAD

**Discussion by AlexanderFarkas** [sqlalchemy/sqlalchemy#12061] [Usage Questions]
I expect entity to be strong-referenced by its back_populates, since I purposefully use 'raise_on_sql' instead of 'raise'.

SQLAlchemy 2.0.36

```python
import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, mapped_column, relationship, Mapped


class Base(DeclarativeBase):
    pass

class Child(Base):
    __tablename__ = "child"
    child_id = mapped_column(sa.Integer, primary_key=True)
    parent_id = mapped_column(sa.Integer, sa.ForeignKey('parent.parent_id'))
    parent: Mapped['Parent'] = relationship(
        back_populates='children',
        lazy="raise_on_sql"
    )

class Parent(Base):
    __tablename__ = "parent"
    parent_id = mapped_column(sa.Integer, primary_key=True)
    children: Mapped[list[Child]] = relationship()

engine = sa.create_engine('sqlite://')
Base.metadata.create_all(engine)
session = sa.orm.Session(engine)

session.add(
    Parent(
        children=[
            Child(),
            Child(),
        ]
    )
)
session.commit()
session.close()

session = sa.orm.Session(engine)
def get_children():
    parent = session.get(Parent, 1)
    # [child.parent for child in parent.children] # Uncomment me to run successfully
    return parent.children

print(get_children()[0].parent) 
```

It throws `sqlalchemy.exc.InvalidRequestError: 'Child.parent' is not available due to lazy='raise_on_sql'`.

If I uncomment this:
```python
 # [child.parent for child in parent.children] # Uncomment me to run successfully
```
it runs successfully.

**Comment by zzzeek** [comment]
the "parent" object is garbage collected because your function is not returning it.    keep it strongly referenced

```py
session = sa.orm.Session(engine)
def get_children():
    parent = session.get(Parent, 1)
    # [child.parent for child in parent.children] # Uncomment me to run successfully
    return parent

parent = get_children()
print(parent.children[0].parent)
```

loading `parent.children` alone is not going to strongly reference `Child.parent`.   Think of the performance implications of this if it did.

**Reply by AlexanderFarkas** [reply]
I understand, but there is no such behaviour if it's "selectin" instead of "raise_on_sql". I would like to keep an object, once it's loaded as part of the previous query. 

The example is very simplified, in the actual program I return another object, which only references children.

**Reply by zzzeek** [reply]
> but there is no such behaviour if it's "selectin" instead of "raise_on_sql".

that's because....you are eagerly loading child.parent.   So yes, "selectin" will eagerly load the object, whereas "raise_on_sql" will not, is expected behavior.

**Reply by AlexanderFarkas** [reply]
It would be nice if it were possible to apply this kind of loading strategy. 
If object is loaded, keep the strong reference to it. If it's not, then raise on SQL
Basically, eager raise_on_sql :)

**Reply by zzzeek** [reply]
you might want to consider "immediateload" here, since that will do the identity map GET operation you are looking for and won't emit sql


```py
session = sa.orm.Session(engine)
def get_children():
    parent = session.get(Parent, 1, options=[defaultload(Parent.children).immediateload(Child.parent)])
    # [child.parent for child in parent.children] # Uncomment me to run successfully
    return parent.children

print(get_children()[0].parent)
```

**Reply by zzzeek** [reply]
> If object is loaded, keep the strong reference to it. 

in your example, `child.parent` is never loaded.    there is no place for a strong reference

**Reply by zzzeek** [reply]
unless you mean the overall behavior of the `Session` as being weak referencing is what you have an issue with.    That wasn't SQLAlchemy's behavior decades ago but eventually became the standard.  you can use [attachment events](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_attach) to reverse this.   I dont have a recipe lying around but I likely have shown this to people in the past

**Reply by zzzeek** [reply]
something like this

```py
@sa.event.listens_for(sa.orm.Session, "after_attach")
@sa.event.listens_for(sa.orm.Session, "loaded_as_persistent")
def strong_ref_everything(session, instance):
    refs = session.info.get("strong_refs", None)
    if refs is None:
        refs = set()
        session.info["strong_refs"] = refs

    refs.add(instance)

session = sa.orm.Session(engine)
def get_children():
    parent = session.get(Parent, 1)
    # [child.parent for child in parent.children] # Uncomment me to run successfully
    return parent.children

print(get_children()[0].parent)
```

**Reply by AlexanderFarkas** [reply]
Thank you for your suggestions!
The nice thing about "raise_on_sql" (on my first impression) was that I could specify it on a Mapper level. 
So, every time I load Child as part of the Parent entity, it doesn't need to issue any SQL. But if I load Child as a single entity in a separate query, I prevent my code from accidentally accessing Parent (since Parent should be used only if it was loaded originally).
